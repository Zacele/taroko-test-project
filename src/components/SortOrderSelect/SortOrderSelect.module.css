.select {
  padding: 8px 16px;
  border: 1px solid #ccc;
  width: 100%;
  @media (min-width: 768px) {
    width: unset;
  }
}
