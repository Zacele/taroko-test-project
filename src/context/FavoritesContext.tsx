'use client';
import React, {
  createContext,
  useContext,
  useState,
  ReactNode,
  useEffect,
} from 'react';
import Cookies from 'js-cookie';

type FavoritesContextType = {
  favorites: number[];
  toggleFavorite: (id: number) => void;
};

const FavoritesContext = createContext<FavoritesContextType | undefined>(
  undefined,
);

type FavoritesProviderProps = {
  children: ReactNode;
};

const FAVORITES_COOKIE_NAME = 'favorites';

export const FavoritesProvider: React.FC<FavoritesProviderProps> = ({
  children,
}) => {
  // Initialize state with empty array or from cookies
  const [favorites, setFavorites] = useState<number[]>(() => {
    const cookieValue = Cookies.get(FAVORITES_COOKIE_NAME);
    return cookieValue ? JSON.parse(cookieValue) : [];
  });

  useEffect(() => {
    // Update cookie whenever favorites change
    Cookies.set(FAVORITES_COOKIE_NAME, JSON.stringify(favorites), {
      expires: 7,
    });
  }, [favorites]);

  const toggleFavorite = (id: number) => {
    setFavorites((currentFavorites) => {
      const isFavorite = currentFavorites.includes(id);
      const newFavorites = isFavorite
        ? currentFavorites.filter((favId) => favId !== id)
        : [...currentFavorites, id];

      return newFavorites;
    });
  };

  return (
    <FavoritesContext.Provider value={{ favorites, toggleFavorite }}>
      {children}
    </FavoritesContext.Provider>
  );
};

export const useFavorites = (): FavoritesContextType => {
  const context = useContext(FavoritesContext);
  if (context === undefined) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};
