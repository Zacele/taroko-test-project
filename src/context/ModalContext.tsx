'use client';
import { ModalState } from '@/interface';
import React, { createContext, useContext, useState, ReactNode } from 'react';

const ModalContext = createContext<ModalState | undefined>(undefined);

export const useModal = () => {
  const context = useContext(ModalContext);
  if (context === undefined) {
    throw new Error('useModal must be used within a ModalProvider');
  }
  return context;
};

export const ModalProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [open, setOpen] = useState(false);
  const [isEditMode, setIsEditMode] = useState(false);
  const [selectedId, setSelectedId] = useState<number | null>(null);

  const closeModal = () => {
    setOpen(false);
    setIsEditMode(false);
    setSelectedId(null);
  };

  const openCreateModal = () => {
    setOpen(true);
  };

  const openEditModal = (id: number) => {
    setIsEditMode(true);
    setOpen(true);
    setSelectedId(id);
  };

  return (
    <ModalContext.Provider
      value={{
        isOpen: open,
        isEdit: isEditMode,
        openModal: openCreateModal,
        closeModal,
        openEditModal,
        selectedId,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};
