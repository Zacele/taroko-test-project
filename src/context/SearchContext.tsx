'use client';
import { ContactType } from '@/interface';
import React, { createContext, useContext, ReactNode, useState } from 'react';

interface SearchContextType {
  searchQuery: string;
  setSearchQuery: (query: string) => void;
  filterContacts: (contacts: ContactType[]) => ContactType[];
}

const SearchContext = createContext<SearchContextType | undefined>(undefined);

interface SearchProviderProps {
  children: ReactNode;
}

export const useSearch = () => {
  const context = useContext(SearchContext);
  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }
  return context;
};

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [searchQuery, setSearchQuery] = useState('');

  const filterContacts = (contacts: ContactType[]): ContactType[] => {
    return contacts.filter((contact) =>
      `${contact.first_name} ${contact.last_name}`
        .toLowerCase()
        .includes(searchQuery.toLowerCase()),
    );
  };

  return (
    <SearchContext.Provider
      value={{ searchQuery, setSearchQuery, filterContacts }}
    >
      {children}
    </SearchContext.Provider>
  );
};
