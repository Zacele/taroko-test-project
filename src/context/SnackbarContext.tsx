'use client';
import React, { createContext, useContext, useState, ReactNode } from 'react';
import Snackbar from '@mui/material/Snackbar';
import Alert, { AlertColor } from '@mui/material/Alert';

type SnackbarContextType = {
  showSnackbar: (message: string, severity: AlertColor) => void;
};

const SnackbarContext = createContext<SnackbarContextType | undefined>(
  undefined,
);

type Props = {
  children: ReactNode;
};

export const SnackbarProvider: React.FC<Props> = ({ children }) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState('');
  const [severity, setSeverity] = useState<AlertColor>('info');

  const showSnackbar = (message: string, severity: AlertColor = 'info') => {
    setMessage(message);
    setSeverity(severity);
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <SnackbarContext.Provider value={{ showSnackbar }}>
      {children}
      <Snackbar open={open} autoHideDuration={2000} onClose={handleClose}>
        <Alert onClose={handleClose} severity={severity} sx={{ width: '100%' }}>
          {message}
        </Alert>
      </Snackbar>
    </SnackbarContext.Provider>
  );
};

export const useSnackbar = () => {
  const context = useContext(SnackbarContext);
  if (context === undefined) {
    throw new Error('useSnackbar must be used within a SnackbarProvider');
  }
  return context;
};
