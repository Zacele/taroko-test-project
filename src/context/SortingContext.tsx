'use client';
import { SortingOrderType } from '@/interface';
import React, { createContext, useContext, useState, ReactNode } from 'react';

interface SortingContextType {
  order: SortingOrderType;
  setOrder: (order: SortingOrderType) => void;
}

const SortingContext = createContext<SortingContextType | undefined>(undefined);

interface SortingProviderProps {
  children: ReactNode;
}

export const SortingProvider: React.FC<SortingProviderProps> = ({
  children,
}) => {
  const [order, setOrder] = useState<SortingOrderType>('desc');

  return (
    <SortingContext.Provider value={{ order, setOrder }}>
      {children}
    </SortingContext.Provider>
  );
};

export const useSortingContext = () => {
  const context = useContext(SortingContext);
  if (context === undefined) {
    throw new Error('useSortingContext must be used within a SortingProvider');
  }
  return context;
};
