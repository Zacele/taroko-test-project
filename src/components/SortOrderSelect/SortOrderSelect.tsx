'use client';

import React from 'react';
import styles from './SortOrderSelect.module.css';
import { useSortingContext } from '@/context/SortingContext';
import { GetContactsResponseType, SortingOrderType } from '@/interface';
import { useQueryClient } from '@tanstack/react-query';
interface Option {
  value: string;
  label: string;
}

const SortOrderSelect: React.FC = () => {
  const { order, setOrder } = useSortingContext();
  const queryClient = useQueryClient();
  const [sortOrder, setSortOrder] = React.useState<SortingOrderType>(order);
  const options: Option[] = [
    { value: 'desc', label: 'A-Z' },
    { value: 'asc', label: 'Z-A' },
  ];

  const handleSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newOrder = event.target.value as SortingOrderType;
    setSortOrder(newOrder);
    setOrder(newOrder);

    queryClient.setQueryData(
      ['contacts'],
      (oldData: GetContactsResponseType) => {
        const sortedData = [...oldData.data].sort((a, b) => {
          const nameA = `${a.first_name} ${a.last_name}`.toLowerCase();
          const nameB = `${b.first_name} ${b.last_name}`.toLowerCase();
          return newOrder === 'desc'
            ? nameA.localeCompare(nameB)
            : nameB.localeCompare(nameA);
        });
        return {
          ...oldData,
          data: sortedData,
        };
      },
    );
  };

  return (
    <select
      value={sortOrder}
      onChange={handleSelectChange}
      className={styles.select}
      defaultValue={sortOrder}
    >
      {options.map((option) => (
        <option key={option.value} value={option.value}>
          {option.label}
        </option>
      ))}
    </select>
  );
};

export default SortOrderSelect;
