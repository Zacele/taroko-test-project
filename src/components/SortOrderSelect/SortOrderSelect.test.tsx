import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import SortOrderSelect from './SortOrderSelect';
import { useSortingContext } from '@/context/SortingContext';
import { useQueryClient } from '@tanstack/react-query';

jest.mock('@/context/SortingContext', () => ({
  useSortingContext: jest.fn(),
}));

jest.mock('@tanstack/react-query', () => ({
  useQueryClient: jest.fn(),
}));

describe('SortOrderSelect', () => {
  beforeEach(() => {
    (useSortingContext as jest.Mock).mockReturnValue({
      order: 'desc',
      setOrder: jest.fn(),
    });

    const queryClientMock = {
      setQueryData: jest.fn(),
    };
    (useQueryClient as jest.Mock).mockReturnValue(queryClientMock);
  });

  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(<SortOrderSelect />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('updates sorting order on selection', async () => {
    const setOrderMock = jest.fn();
    (useSortingContext as jest.Mock).mockReturnValue({
      order: 'desc',
      setOrder: setOrderMock,
    });

    render(<SortOrderSelect />);

    await userEvent.selectOptions(screen.getByRole('combobox'), ['asc']);
    expect(setOrderMock).toHaveBeenCalledWith('asc');
  });
});
