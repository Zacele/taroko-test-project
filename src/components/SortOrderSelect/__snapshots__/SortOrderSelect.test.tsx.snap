// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortOrderSelect renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <select
    class="select"
  >
    <option
      value="desc"
    >
      A-Z
    </option>
    <option
      value="asc"
    >
      Z-A
    </option>
  </select>
</DocumentFragment>
`;
